optaplanner-core/src/main/java/org/optaplanner/core/api/solver/SolverManager.java

```diff
@@ -0,0 +1,203 @@
+            Function<ProblemId_, Solution_> problemFinder, Consumer<Solution_> bestSolutionConsumer) {
+        return solveObserving(problemId, problemFinder, bestSolutionConsumer, null);
+    }
+
+    /**
+     * As defined by {@link #solveObserving(Object, Function, Consumer)}.
+     * @param problemId never null, a ID for each planning problem. This must be unique.
+     * Use this problemId to {@link #terminateEarly(Object) terminate} the solver early,
+     * {@link #getSolverStatus(Object) to get the status} or if the problem changes while solving.
+     * @param problemFinder never null, function that returns a {@link PlanningSolution}, usually with uninitialized planning variables
+     * @param bestSolutionConsumer never null, called multiple times, on a consumer thread
+     * @param exceptionHandler sometimes null, called if an exception or error occurs.
+     * If null it defaults to logging the exception as an error.
+     * @return never null
+     */
+    SolverJob<Solution_, ProblemId_> solveObserving(ProblemId_ problemId,
+            Function<ProblemId_, Solution_> problemFinder, Consumer<Solution_> bestSolutionConsumer,
+            BiConsumer<ProblemId_, Throwable> exceptionHandler);
+
+    /**
+     * Returns if the {@link Solver} is scheduled to solve, actively solving or not.
+     * <p>
+     * Returns {@link SolverStatus#NOT_SOLVING} if the solver already terminated or if the problemId was never added.
+     * To distinguish between both cases, use {@link SolverJob#getSolverStatus()} instead.
+     * Here, that distinction is not supported because it would cause a memory leak.
+     * @param problemId never null, a value given to {@link #solveBatch(Object, Function, Consumer)}
+     * or {@link #solveObserving(Object, Function, Consumer)}
+     * @return never null
+     */
+    SolverStatus getSolverStatus(ProblemId_ problemId);
+
+    // TODO Future features
+//    void reloadProblem(ProblemId_ problemId, Function<ProblemId_, Solution_> problemFinder);
+
+    // TODO Future features
+//    void addProblemFactChange(ProblemId_ problemId, ProblemFactChange<Solution_> problemFactChange);
+
+    /**
+     * Terminates the solver or cancels the solver job if it hasn't (re)started yet.
+     * <p>
+     * Does nothing if the solver already terminated or the problemId was never added.
+     * To distinguish between both cases, use {@link SolverJob#terminateEarly()} instead.
+     * Here, that distinction is not supported because it would cause a memory leak.
+     * @param problemId never null, a value given to {@link #solveBatch(Object, Function, Consumer)}
+     * or {@link #solveObserving(Object, Function, Consumer)}
+     */
+    void terminateEarly(ProblemId_ problemId);
+
+    /**
+     * Calculates the {@link Score} of a {@link PlanningSolution} and updates its {@link PlanningScore} member.
+     * @param solution never null
+     */
+    void updateScore(Solution_ solution);
```

Comment:
This is surprising to me. Why give users the power to do this? This looks like a call that would be internal to the solver. It completely breaks the abstraction of this class.

Reply:
Users need it when loading a timetable from the database and showing it in the UI. It's already solved, but the UI needs to show the score (and indictments later).
Let's discuss alternative proposals for this tomorrow .

Reply:
Agreed to move into ScoreService (a new class)